addons/l10n_vn_edi/models/sinvoice.py

```diff
@@ -0,0 +1,129 @@
+                ('l10n_vn_edi_state', 'not in', ('ready_to_send', False)),  # Only matches sent invoices.
+            ],
+            groupby=['l10n_vn_edi_invoice_symbol'],
+            aggregates=['__count'],
+        )
+        invoices_per_template = defaultdict(int)
+        for symbol, count in invoice_counts:
+            invoices_per_template[symbol.invoice_template_id.id] += count
+
+        for record in self:
+            if invoices_per_template[record.id] > 0:
+                raise UserError(_('You cannot change the code or type of the template %s because it has '
+                                  'already been used to send invoices.', record.name))
+
+
+# Invoice symbol that needs to be passed to Sinvoice and will determine the prefix of the
+# invoice number on their system
+class SInvoiceSymbol(models.Model):
+    _name = 'sinvoice.symbol'
+    _description = 'SInvoice symbol'
+    """
+    The invoice symbols are made of multiple parts.
+    The symbols characters have meanings that will influence the resulting invoices:
+
+    The first character is either C (invoice with code from tax department or K (without)
+    It is then followed by two number which represent the current year (24 in 2024)
+    Next, a single character representing additional invoice information:
+
+        T: E-invoices registered with tax authorities by enterprises, organizations or household businesses;
+        D: E-invoices used for sale of public property or national reserve goods or special e-invoices that do not have some mandatory contents of the “T” invoices;
+        L: E-invoices issued separately by tax authorities;
+        M: E-invoices generated by cash registers;
+        N: Electronic delivery and internal consignment notes;
+        B: Electronic delivery notes for goods sent to sales agents;
+        G: VAT invoices in the form of electronic stamps, tickets or cards;
+        H: Sales invoices in the form of electronic stamps, tickets or cards.
+
+    Finally, it ends with two characters that shall be decided by the seller for their management requirements.
+    """
+
+    name = fields.Char(
+        string='Symbol',
+        required=True,
+    )
+    invoice_template_id = fields.Many2one(
+        comodel_name='sinvoice.template',
+        required=True,
+    )
+
+    _sql_constraints = [
+        ('name_template_uniq', 'unique (name, invoice_template_id)', 'The combination symbol/template must be unique!')
+    ]
+
+    @api.constrains('name', 'invoice_template_id')
+    def _constrains_changes(self):
+        """
+        Multiple API endpoints will use these data, we should thus not allow changing them if they have been used
+        for any invoices sent to sinvoice.
+        """
+        invoice_counts = self.env['account.move']._read_group(
+            domain=[
+                ('l10n_vn_edi_invoice_symbol.invoice_template_id', 'in', self.ids),
```

Comment:
Oupsie

We can't call the template constraint from the symbol since you could have two symbols A & B using the same template.
If A is being used but not B, you still want to allow B to be edited.

On the other hand, I suppose the template could call the constrains on invoice_symbols_ids so that it is raised if any of them is being used :+1: